src/dmd/todt.d

```diff
@@ -211,378 +211,383 @@ extern (C++) void Initializer_toDt(Initializer init, ref DtBuilder dtb)
-                            // Pad remainder with 0
-                            dtb.nzeros(cast(uint)((dim - n) * tsa.next.size()));
-                        }
-                    }
-                    break;
-                }
+    void visitString(StringExp e)
+    {
+        //printf("StringExp.toDt() '%s', type = %s\n", e.toChars(), e.type.toChars());
+        Type t = e.type.toBasetype();
 
-                default:
-                    printf("StringExp.toDt(type = %s)\n", e.type.toChars());
-                    assert(0);
-            }
-            if (p != e.toPtr())
-                mem.xfree(p);
+        // BUG: should implement some form of static string pooling
+        int n = cast(int)e.numberOfCodeUnits();
+        char* p = e.toPtr();
+        if (!p)
+        {
+            p = cast(char*)mem.xmalloc(n * e.sz);
+            e.writeTo(p, false);
         }
 
-        override void visit(ArrayLiteralExp e)
+        switch (t.ty)
         {
-            //printf("ArrayLiteralExp.toDt() '%s', type = %s\n", e.toChars(), e.type.toChars());
+            case Tarray:
+                dtb.size(n);
+                goto case Tpointer;
 
-            auto dtbarray = DtBuilder(0);
-            for (size_t i = 0; i < e.elements.dim; i++)
-            {
-                Expression_toDt(e.getElement(i), dtbarray);
-            }
+            case Tpointer:
+                if (e.sz == 1)
+                {
+                    import dmd.e2ir : toStringSymbol;
+                    import dmd.glue : totym;
+                    Symbol* s = toStringSymbol(p, n, e.sz);
+                    dtb.xoff(s, 0);
+                }
+                else
+                    dtb.abytes(0, n * e.sz, p, cast(uint)e.sz);
+                break;
 
-            Type t = e.type.toBasetype();
-            switch (t.ty)
+            case Tsarray:
             {
-                case Tsarray:
-                    dtb.cat(dtbarray);
-                    break;
+                TypeSArray tsa = cast(TypeSArray)t;
 
-                case Tpointer:
-                case Tarray:
+                dtb.nbytes(n * e.sz, p);
+                if (tsa.dim)
                 {
-                    if (t.ty == Tarray)
-                        dtb.size(e.elements.dim);
-                    dt_t* d = dtbarray.finish();
-                    if (d)
-                        dtb.dtoff(d, 0);
-                    else
-                        dtb.size(0);
-
-                    break;
+                    dinteger_t dim = tsa.dim.toInteger();
+                    if (n < dim)
+                    {
+                        // Pad remainder with 0
+                        dtb.nzeros(cast(uint)((dim - n) * tsa.next.size()));
+                    }
                 }
-
-                default:
-                    assert(0);
+                break;
             }
+
+            default:
+                printf("StringExp.toDt(type = %s)\n", e.type.toChars());
```

Comment:
maybe comment this ?

Reply:
This PR only does one thing, that's replace Visitor with switch.

Reply:
sorry i reviewed too fast and only with the diff.

Reply:
yeah you could leave it, my bad, I missed the assert(0) beyound.